### PR TITLE
RAIL-4168 Add support for change height in edit mode

### PR DESCRIPTION
**Pull request by MilanLempera**
- move grid Row to grid-row level
  - add render mode prop to layout components
  - add height hotspot and preview
  - add command and handler for height change
  - add scrolling when dragging above and below

JIRA: RAIL-4168

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8770/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10609/ : SUCCESS in 21m 56s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8777/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10619/ : SUCCESS in 19m 00s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8778/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10620/ : SUCCESS in 12m 25s

File: libs/sdk-ui-dashboard/src/model/commands/layout.ts [hunk 1]
```diff
@@ -807,3 +808,64 @@ export function revertLastLayoutChange(correlationId?: string): UndoLayoutChange
         payload: {},
     };
 }
+
+/**
+ * Payload of the {@link AddLayoutSection} command.
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commands/layout.ts:813]: Mismatching TSDoc

File: libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+) {
+    const {
+        payload: { sectionIndex, itemIndexes },
+    } = command;
+
+    if (!validateSectionExists(layout, sectionIndex)) {
+        throw invalidArgumentsProvided(
+            ctx,
+            command,
+            `Attempting to replace item from non-existent section at ${sectionIndex}. There are only ${layout.sections.length} sections.`,
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts:33]: Should probably be resize, not replace?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    const fromSection = layout.sections[sectionIndex];
+
+    itemIndexes.forEach((itemIndex) => {
+        if (!validateItemExists(fromSection, itemIndex)) {
+            throw invalidArgumentsProvided(
+                ctx,
+                command,
+                `Attempting to replace non-existent item from index ${itemIndex} in section ${sectionIndex}. There are only ${fromSection.items.length} items in this section.`,
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts:44]: Should probably be resize, not replace?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    const layout = yield select(selectLayout);
+    const insightsMap = yield select(selectInsightsMap);
+
+    validateLayoutIndexes(ctx, layout, cmd);
+
+    validateHeight(ctx, layout, insightsMap, cmd);
+
+    yield put(
+        batchActions([
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts:66]: Do you need to batch the actions if there is only one?

File: libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    const minLimit = getMinHeight(widgets, insightsMap);
+    const maxLimit = getMaxHeight(widgets, insightsMap);
+
+    const validHeight = height >= minLimit && height <= maxLimit;
+
+    if (!validHeight) {
+        throw invalidArgumentsProvided(
+            ctx,
+            cmd,
+            `Attempting to set invalid height. Allowed height is form ${minLimit} to ${maxLimit}.`,
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/layout/resizeHeightHandler.ts:101]: Typo: from

File: libs/sdk-ui-dashboard/src/model/commands/layout.ts [hunk 6]
```diff
@@ -807,3 +808,64 @@ export function revertLastLayoutChange(correlationId?: string): UndoLayoutChange
...
+    readonly type: "GDC.DASH/CMD.FLUID_LAYOUT.RESIZE_HEIGHT";
+    readonly payload: ResizeHeightPayload;
+}
+
+/**
+ * Creates the ResizeHeight command.
+ *
+ * @param sectionIndex - index of the section
+ * @param itemIndexes - indexes of the items
+ * @param height - height in GP
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/commands/layout.ts:850]: What does GP mean? Can you elaborate in the TSDoc please?

File: libs/sdk-ui-dashboard/src/model/layout/sizing.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        return INSIGHT_WIDGET_SIZE_INFO_DEFAULT;
+    }
+    return sizeInfo;
+};
+
+const getKpiSizeInfo = (settings: ISettings, kpi?: MeasurableWidgetContent): IVisualizationSizeInfo => {
+    if (!settings.enableKDWidgetCustomHeight) {
+        return KPI_WIDGET_SIZE_INFO_DEFAULT_LEGACY;
+    }
+    if (!isKpi(kpi)) {
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/model/layout/sizing.ts:67]: Why is this condition here? isKpi can be false in getKpiSizeInfo?

Comment (no23reason) [reply libs/sdk-ui-dashboard/src/model/layout/sizing.ts:67]: Or is it for placeholders?

File: libs/sdk-ui-dashboard/src/presentation/dragAndDrop/DragLayerPreview/HeightResizerDragPreview.tsx [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    const currentUnlimitedHeightGR = getNewHeightGR(
+        item.widgetHeights,
+        currentOffsetY,
+        documentDimensions.scrollTop,
+        item.initialScrollTop,
+    );
+    const hasNowReachedLimit = hasHeightReachedLimit(currentUnlimitedHeightGR, item.minLimit, item.maxLimit);
+
+    if (hasNowReachedLimit !== hasReachedLimit) {
+        setReachedLimit(hasNowReachedLimit);
```
Comment (no23reason) [inline libs/sdk-ui-dashboard/src/presentation/dragAndDrop/DragLayerPreview/HeightResizerDragPreview.tsx:33]: This is suspicious: a side effect in render. Should this be an useEffect instead?

File: libs/sdk-ui-dashboard/src/model/layout/sizing.ts (hunk 7 above)
Comment (MilanLempera) [reply libs/sdk-ui-dashboard/src/model/layout/sizing.ts:67]: indeed, I have no idea, this function is moved from layout sizing utils without changes https://github.com/gooddata/gooddata-ui-sdk/pull/2655/files#diff-2fab94c22a67fb77aa71b00ad301467efa1f26c437a974e82c2a364b8e7854f2L484

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8786/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10630/ : FAILURE in 5m 04s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8789/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10633/ : SUCCESS in 10m 56s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result
 - 1 VULNERABILITY

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2655
The scan job will fail because a vulnerability has been found

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8793/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10639/ : FAILURE in 5m 40s

**Comment by yenkins** [comment]
### Sonar scan result
 - 1 VULNERABILITY

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-check-extended-sonar-PR2655
The scan job will fail because a vulnerability has been found

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by yenkins** [comment]
### Sonar scan result
 - 1 VULNERABILITY

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2655
The scan job will fail because a vulnerability has been found

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by yenkins** [comment]
### Sonar scan result
 - 1 VULNERABILITY

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2655
The scan job will fail because a vulnerability has been found

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2519/ : SUCCESS in 3m 10s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2818/ : SUCCESS in 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1644/ : FAILURE in 18m 18s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10647/ : SUCCESS in 10m 44s

**Comment by MilanLempera** [comment]
extended check sonar

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8799/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1645/ : FAILURE in 18m 38s

**Comment by MilanLempera** [comment]
extended check sonar

**Comment by yenkins** [comment]
### Sonar scan result
 - 1 VULNERABILITY

More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-check-extended-sonar-PR2655
The scan job will fail because a vulnerability has been found

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8800/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1646/ : FAILURE in 17m 31s

**Comment by MilanLempera** [comment]
extended check sonar

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8811/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10658/ : SUCCESS in 11m 20s

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-check-extended-sonar-PR2655

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8812/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1649/ : SUCCESS in 17m 20s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/8813/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10659/ : SUCCESS in 11m 15s

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR2655

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/2522/ : SUCCESS in 2m 46s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/2821/ : SUCCESS in 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/1650/ : SUCCESS in 17m 50s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/10660/ : SUCCESS in 12m 49s

**Pull request merged**